micrometer-core/src/main/java/io/micronaut/configuration/metrics/micrometer/intercept/TimedInterceptor.java

```diff
@@ -75,86 +79,123 @@ protected TimedInterceptor(MeterRegistry meterRegistry) {
+                                metricName,
+                                sample.get(),
+                                throwable.getClass().getSimpleName(),
+                                metadata))
+                        .doOnSuccess(o -> stopTimed(
                                 metricName,
-                                sample,
-                                throwable != null ? throwable.getClass().getSimpleName() : "none",
+                                sample.get(),
+                                "none",
                                 metadata));
-                    } else {
-                        return result;
-                    }
-                } catch (Exception e) {
-                    exceptionClass = e.getClass().getSimpleName();
-                    throw e;
-                } finally {
-                    if (stop) {
-
-                        stopTimed(metricName, sample, exceptionClass, metadata);
-                    }
-                }
+
+                return Publishers.convertPublisher(single, javaReturnType);
+            } else {
+                AtomicReference<String> exceptionClass = new AtomicReference<>("none");
+                Flowable<?> flowable = Publishers.convertPublisher(result, Flowable.class);
+                flowable = flowable.doOnRequest(n -> sample.set(Timer.start(meterRegistry)))
+                        .doOnError(throwable -> exceptionClass.set(throwable.getClass().getSimpleName()))
+                        .doOnComplete((Action) () -> {
+                            final Timer.Sample s = sample.get();
+                            if (s != null) {
+                                stopTimed(
+                                        metricName,
+                                        s,
+                                        exceptionClass.get(),
+                                        metadata);
+                            }
+                        });
+                return Publishers.convertPublisher(flowable, javaReturnType);
             }
         }
-        return context.proceed();
+    }
+
+    private Object handleKotlinCoroutine(MethodInvocationContext<Object, Object> context, final AnnotationMetadata metadata, final String metricName) {
+        final Timer.Sample sample = Timer.start(meterRegistry);
+        final Object result = context.proceed();
+        // is it suspended?
+        if (result == null || !result.getClass().getName().equals("kotlin.coroutines.intrinsics.CoroutineSingletons")) {
+            // not suspended
+            stopTimed(
+                    metricName,
+                    sample,
+                    "none",
+                    metadata);
+        } else {
+            // last argument
+            final String[] argumentNames = context.getArgumentNames();
+            final MutableArgumentValue<?> argument = context.getParameters().get(argumentNames[argumentNames.length - 1]);
```

Comment:
`context.getParameters()` is relatively expensive, and you don't need the `MutableArgument` value for these case. Use `getParameterValues()[argumentNames.length - 1]` instead which will provide better performance

Reply:
Done